cmd/expired-authz-purger/main.go

```diff
@@ -46,79 +43,61 @@ type expiredAuthzPurger struct {
 }
 
 // purge looks up pending or finalized authzs (depending on the value of
-// `table`) that expire before `purgeBefore`. If `yes` is true, or if a user at
-// the terminal types "y", it will then delete those authzs, using `parallelism`
+// `table`) that expire before `purgeBefore`, using `parallelism`
 // goroutines. It will delete a maximum of `max` authzs.
 // Neither table has an index on `expires` by itself, so we just iterate through
 // the table with LIMIT and OFFSET using the default ordering. Note that this
 // becomes expensive once the earliest set of authzs has been purged, since the
 // database will have to scan through many rows before it finds some that meet
 // the expiration criteria. When we move to better authz storage (#2620), we
 // will get an appropriate index that will make this cheaper.
-func (p *expiredAuthzPurger) purge(table string, yes bool, purgeBefore time.Time, parallelism int, max int) error {
-	var ids []string
-	for len(ids) < max {
-		var idBatch []string
-		var query string
-		switch table {
-		case "pendingAuthorizations":
-			query = "SELECT id FROM pendingAuthorizations WHERE expires <= ? ORDER BY id LIMIT ? OFFSET ?"
-		case "authz":
-			query = "SELECT id FROM authz WHERE expires <= ? ORDER BY id LIMIT ? OFFSET ?"
-		}
-		_, err := p.db.Select(
-			&idBatch,
-			query,
-			purgeBefore,
-			p.batchSize,
-			len(ids),
-		)
-		if err != nil && err != sql.ErrNoRows {
-			return err
-		}
-		if len(idBatch) == 0 {
-			break
-		}
-		ids = append(ids, idBatch...)
-	}
-	if len(ids) > max {
-		ids = ids[:max]
+func (p *expiredAuthzPurger) purge(table string, purgeBefore time.Time, parallelism int, max int) error {
+	var query string
+	switch table {
+	case "pendingAuthorizations":
```

Comment:
It seems like these two queries are the same except for the table name - would it be simpler to have one `query` and `sprintf` in the table name?

Reply:
I think we discussed this when originally landing this code. I really want to avoid multiple levels of interpolation into SQL strings, in particular generating a string with sprintf to then be processed for `?` and `:foo` by the SQL engine. Even though it's clearly safe in this case, it takes a little cognitive overhead each time you review it to double check. Or in other words, it's a code smell. In this case I think the code smell of duplication is less stinky. :-)

Reply:
That makes sense. It's not a big sticking point for me & I agree we should consider raw interpolation smelly :-)